WalletWasabi/WabiSabi/Models/DecompositionAlgs/GreedyDecomposer.cs

```diff
@@ -0,0 +1,67 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.DecompositionAlgs
+{
+	public class GreedyDecomposer
+	{
+		public GreedyDecomposer(IEnumerable<Money> preferredDenominations, Money dustThreshold, FeeRate feeRate)
+		{
+			Decomposition = new(preferredDenominations);
+			DustThreshold = dustThreshold;
+			FeeRate = feeRate;
+		}
+
+		public Decomposition Decomposition { get; }
+		private Money DustThreshold { get; }
+		private FeeRate FeeRate { get; }
+
+		public void Decompose(Coin coin)
+		{
+			Money remaining = coin.Amount - FeeRate.GetFee(coin.ScriptPubKey.EstimateOutputVsize());
+
+			while (remaining > DustThreshold)
+			{
+				Money? denom = LargestDenomBelowIncl(remaining);
+				if (denom is null || denom == Money.Zero)
+				{
+					break;
+				}
+				Decomposition.AddSorted(denom);
+
+				var effectiveCost = coin.Amount + FeeRate.GetFee(coin.ScriptPubKey.EstimateOutputVsize());
+
+				remaining -= effectiveCost;
+			}
+
+			if (remaining > DustThreshold)
+			{
+				Decomposition.AddSorted(remaining);
+			}
+		}
+
+		private Money? LargestDenomBelowIncl(Money amount)
```

Comment:
I really hate `null`s. Could we return `Money.Zero` or rename the method to `TryGetLagestDenominationBlahBlah`?

Reply:
Fixed.